ZScript abstract functions

Player701 opened the pull request:
**This PR implements a functionality to define abstract functions in ZScript.**

Example syntax:

```
class Base abstract
{
    abstract void Func();
}

class Derived : Base
{
    override void Func()
    {
        //...
    }
}
```

Notes:

* Abstract functions are implicitly virtual, and cannot be marked virtual or override.
* Abstract functions cannot have a body.
* Abstract functions can only be defined in an abstract class.
* Any non-abstract class must override all parent abstract functions.

### Justification

The primary justification for abstract functions is that they can help reduce the amount of potential bugs in user code by checking for errors at compile time. If a base class has virtual functions that are always overridden in subclasses, it is easy to forget to override one or two of them. A workaround for now is to put calls to ```ThrowAbortException``` into these virtuals so that GZDoom aborts with a meaningful error message indicating a programming error. However, this necessitates writing all this extra code by hand, while the best solution would be to simply make the base functions abstract.

### Technical information

The implementation is rather simple, although we do need some safety checks mostly to ensure that abstract classes can never be instantiated.

* First of all, I've extended the check in ```PClass::CreateNew```. This is a fallback case for all instances where non-actor objects are created by the engine. There seem to be quite a few of those, for example, StatusBarClass in MAPINFO, event handlers, menus and so on - there may be more, and this is the most efficient way to handle them all at once.
* I've also moved the check for ```bAbstract``` from the summon CCMD handler to ```AActor::StaticSpawn```. In this case, we only output a warning instead of an error. This covers all cases where actors get spawned via DECORATE, ZScript and ACS - since this is likely to happen with user code, we should try to be user-friendly and do not error out to the console. However, in case Graf or other devs consider a simple warning not enough, I can make this an error instead and reinstate the original check in the summon handler.
* Finally, we should check that we are not calling an abstract function in ```VMFunction::FirstScriptCall``` - although this should never happen normally, it may indicate a bug, and we should error out gracefully instead of crashing so that it will be easier to investigate the issue.

Note that since it's not possible to override functions in DECORATE, I've also added a check to ```DecoDerivedClass``` so that an actor class with abstract functions does not get created.

### Unit tests

I've written a simple PowerShell-based unit test suite that allows loading DECORATE and ZScript files and check for presence or absence of certain error messages in the log. This way, we can ensure that everything works as intended. I'm sharing the test suite [here on my personal server](https://files.player701.ru/abstract_tests.7z). Simply extract all files to a separate subfolder and run ```powershell ./test.ps1 -gzdoom_path <path_to_your_gzdoom_exe>```.

Optional arguments:

* ```-gzdoom_iwad```: if you don't have doom2.wad then specify another IWAD
* ```-gzdoom_run_delay```: how much to wait for GZDoom to complete startup (default: 5 seconds)
* ```-gzdoom_stop_delay```: how much to wait after GZDoom process has been killed - for some reason it doesn't release locks on temporary files right away (default: 1 second)

Test away! And hey, remember that the test system may have bugs since I've spent basically only a couple hours on it. In case you have more tests to propose, go ahead. If you think this PR should be tested more, then let us take all the time we need.

Player701 commented:
Oops, looks like some CI builds failed. I've fixed the error and force-pushed, let's see if it builds now.

MekBoss commented:
> Abstract functions can only be defined in an abstract class.

Its an overkill IMO.
Allowing creation of abstract functions only in purely abstract classes is too much...unless abstract class now produce an error if Gzdoom tries to initialize such class.

Player701 commented:
In the current version, GZDoom will throw a compiler error if abstract classes are instantiated with the ```new``` operator in script code. Additionally, it checks before summoning an actor that its class is not marked abstract. However, it still allows instantiating these classes by calling ```Actor.Spawn``` and by multiple other methods, such as via ACS and DECORATE/ZScript functions ```A_SpawnItemEx``` et al. This is technically not correct, and my PR addresses these issues.

Note that programming languages like C# and Java work exactly the same way - abstract methods can only be defined in abstract classes. This also makes it easier to verify that we aren't trying to instantiate a class that contains such methods by checking for a single flag rather than iterating through all of its virtuals.

coelckers commented:
In C++ defining an abstract function makes the entire class abstract by default, but of course it comes with the problem that a class can become abstract by accident which is also not that great.

Player701 commented:
Agreed. In this case, explicit is better than implicit.

Player701 commented:
Epic! Thanks a lot!
